Constraints to state variable handling by integrator's schemes

dionperd opened the pull request:
Followed clamp_state example, but took if statements outside clamp_state and constraint_state methods to avoid the function stack when no constraint or clamping applies.

In the base model I am checking for consistency of the user input or model property on constraints.

I updated the two wong-wang models by adding the constraints and modifying the _numba_dfun of the old model, to not have a reflective boundary any more.

In the simulator, I configure the integrator when then model has constraints, and I constrain the history during history configuration.

I added a test similar to the test for state clamping.

dionperd committed:
Added constraints handling. Version 1. Not tested.

dionperd committed:
Added constraints checking back into the dfun of the RedWongWang
to be independent of constraints checked in the integrator's scheme.

dionperd committed:
Debugged and tested changes to simulator

dionperd committed:
Debugged and tested changes to simulator

dionperd committed:
Changes in imports in tvb/basic/traits/__init__.py.

dionperd committed:
Got the if statements outside state clamp and constrain methods in integrator

dionperd committed:
Added if statement outside state constrain of history

dionperd committed:
Removed TODOs on constraints from red wong wang model

dionperd committed:
Converted reflective to saturated constraints for wong-wang _numba_dfun

maedoc commented on tvb/simulator/integrators.py:
```diff
@@ -88,6 +88,17 @@ class Integrator(core.Type):
         it is consitent with Monitors using sample periods corresponding to
         powers of 2 from 128 to 4096Hz.""")
 
+    constraint_state_variable_indices = arrays.IntegerArray(
```
constrained or bounded would be better choice than constraint

dionperd replied:
Done!

maedoc commented on tvb/simulator/integrators.py:
```diff
@@ -109,13 +119,24 @@ def scheme(self, X, dfun, coupling, local_coupling, stimulus):
 
+    def constrain_state(self, X):
+        for sv_ind, sv_bounds in \
+                zip(self.constraint_state_variable_indices,
```
a SIMD friendly form will do better performance wise
```python
vi, vb = self.constraint_state_variable_indices, self.constraint_state_variable_bounds
X[vi] = X[vi] + vb[:, 0]*(X[vi] < vb[:, 0]) + vb[:, 1]*(X[vi] > vb[:, 1])
```
but a numba dfun will be ideal to avoid temporaries.

dionperd replied:
I did something similar removing the temp. But, mind that for one sided boundaries, the other side contains a None value + I got an error for the different arrays sizes in the addition there, because x[vi] is of size space x modes... So, I thought it is getting more complicated...

maedoc commented on tvb/simulator/integrators.py:
```diff
@@ -177,13 +198,19 @@ def scheme(self, X, dfun, coupling, local_coupling, stimulus):
-        inter = X + self.dt * (m_dx_tn  + stimulus)
-        self.clamp_state(inter)
+        inter = X + self.dt * (m_dx_tn + stimulus)
+        if self.constraint_state_variable_boundaries is not None:
```
this is a lot of boilerplate, should be refactored into the call site in the simulator

maedoc commented on tvb/simulator/models/wong_wang.py:
```diff
@@ -39,13 +39,15 @@ def _numba_dfun(S, c, a, b, d, g, ts, w, j, io, dx):
-        x = w[0]*j[0]*S[0] + io[0] + j[0]*c[0]
-        h = (a[0]*x - b[0]) / (1 - numpy.exp(-d[0]*(a[0]*x - b[0])))
-        dx[0] = - (S[0] / ts[0]) + (1.0 - S[0]) * h * g[0]
+        S_constraint = S[0]
```
this could again be a SIMD-friendly
```python
S_c = S[0]*(S[0] > 0.0)*(S[0] < 1.0) + (S[0] < 0.0) + (S[0] > 1.0)
```
(but this may not be a big deal performance wise considering the rest of the simulator

maedoc replied:
but then again, I don't understand why you didn't just remove this code here: if you've already constrained S after the integration step, it should always lie on the interval [0, 1].

dionperd replied:
You are write. I removed them.

maedoc commented on tvb/tests/library/simulator/integrators_test.py:
```diff
@@ -132,6 +132,30 @@ def test_scipy_dop853(self):
         self._scipy_scheme_tester('Dopri5')
 
+    def test_constrain(self):
+        vode = integrators.VODE(
```
why test with VODE?  what do you use in your multiscale stuff?

dionperd replied:
Simpler... Clearer...?

dionperd replied:
Just simpler I guess.

maedoc replied:
I meant why VODE instead of say Euler. But it's OK this doesn't need to change.

maedoc commented:
hi @dionperd good job, ok for me but consider the comments a bit.  the only part I would insist on is changing constraint to bounded or constrained, since it will become part of API and can't change later.

dionperd committed:
Changed names form constraints to bound...
Removed boundary check inside wong-wang models
Removed temp from bound_state method.

liadomide commented:
@maedoc Will you merge this, or should I do it? 
I could migrate it further, into neotrait-py3 branch also, tvb-root and SVN, and make sure the unit-test written by Denis work in all of these.

maedoc commented:
It doesn't seem to interact with neotraits or py3 so I would rebase on trunk and sync with svn as usual, via the tvb-library repo, as the PR is here.

If you'd like, I can do the rebase/merge and let you do the svn sync?

liadomide commented:
yes. Pls do as you propose
